Client/Frontend/Home/PanelDataObservers.swift

```diff
@@ -47,29 +47,42 @@ class ActivityStreamDataObserver: DataObserver {
     deinit {
         events.forEach { NotificationCenter.default.removeObserver(self, name: $0, object: nil) }
     }
-    
-    func invalidate(highlights: Bool) {
-        self.delegate?.willInvalidateDataSources()
 
-        let notify = {
-            self.delegate?.didInvalidateDataSources()
+    /*
+     refreshIfNeeded will refresh the underlying caches for both Topsites and Highlights.
+     By default this will only refresh the highlights if the last fetch is older than 15 mins
+     By default this will only refresh topsites if KeyTopSitesCacheIsValid is false
+     */
+    func refreshIfNeeded(forceHighlights highlights: Bool, forceTopSites topSites: Bool) {
+        guard !profile.isShutdown else {
+            return
         }
-        
-        let invalidateTopSites: () -> Success = {
-            self.profile.history.setTopSitesNeedsInvalidation()
-            return self.profile.history.updateTopSitesCacheIfInvalidated() >>> succeed
+
+        // Highlights are cached for 15 mins 200 - 0 > 900 || 200 < 900
+        let invalidateHighlights = highlights ? true : (Timestamp.uptimeInMilliseconds() - lastInvalidation > invalidationTime)
+        lastInvalidation = invalidateHighlights ? Timestamp.uptimeInMilliseconds() : lastInvalidation
+
+        // KeyTopSitesCacheIsValid is false when we want to invalidate. Thats why this logic is so backwards
+        let invalidateTopSites = topSites ? true : !(profile.prefs.boolForKey(PrefsKeys.KeyTopSitesCacheIsValid) ?? false)
+        if !invalidateTopSites && !invalidateHighlights {
+            // There is nothing to refresh. Bye
+            return
         }
 
-        let shouldInvalidate = highlights ? true : (Date.now() - lastInvalidation > invalidationTime)
-        lastInvalidation = shouldInvalidate ? Date.now() : lastInvalidation
-        let query = shouldInvalidate ? [self.profile.recommendations.invalidateHighlights, invalidateTopSites] : [invalidateTopSites]
-        accumulate(query) >>> effect(notify)
+        self.delegate?.willInvalidateDataSources()
+        self.profile.recommendations.repopulateAll(invalidateTopSites, invalidateHighlights: invalidateHighlights).uponQueue(.main) { _ in
+            if invalidateTopSites {
+                self.profile.prefs.setBool(true, forKey: PrefsKeys.KeyTopSitesCacheIsValid)
+            }
+            self.delegate?.didInvalidateDataSources()
+        }
     }
-    
+
     @objc func notificationReceived(_ notification: Notification) {
         switch notification.name {
         case NotificationProfileDidFinishSyncing, NotificationFirefoxAccountChanged, NotificationPrivateDataClearedHistory:
-            invalidate(highlights: true)
+            profile.prefs.setBool(false, forKey: PrefsKeys.KeyTopSitesCacheIsValid)
```

Comment:
Why bother setting this pref when you're immediately calling with `forceTopSites: true`?